dummyfs/dummyfs.c

```diff
@@ -511,470 +639,509 @@ int dummyfs_create(void *ctx, oid_t *dir, const char *name, oid_t *oid, unsigned
+			o->data = dummyfs_calloc(fs, o->size);
+			if (o->data == NULL) {
+				(void)_dummyfs_truncateObject(fs, o, oldsz);
+				return -ENOMEM;
+			}
 		}
-		else {
-			memset(buff, 0, readsz);
+		memcpy((char *)o->data + offs, buff, len);
+		cnt = len;
+	}
+	else {
+		/* Big file */
+		if (o->chunks == NULL) {
+			size_t tabsz = sizeof(void *) * DUMMYFS_CHUNKCNT(o->size);
+			o->chunks = dummyfs_calloc(fs, tabsz);
+			if (o->chunks == NULL) {
+				(void)_dummyfs_truncateObject(fs, o, oldsz);
+				return -ENOMEM;
+			}
 		}
 
-		len -= readsz;
-		buff += readsz;
-		offs += readsz;
-		ret += readsz;
+		/* Preallocate chunks to avoid ENOMEM on partial write */
+		size_t limit = DUMMYFS_CHUNKIDX(offs + len - 1);
+		for (size_t i = DUMMYFS_CHUNKIDX(offs); i <= limit; ++i) {
+			if (o->chunks[i] == NULL) {
+				o->chunks[i] = dummyfs_mmap(fs);
+				if (o->chunks[i] == NULL) {
+					(void)_dummyfs_truncateObject(fs, o, oldsz);
+					return -ENOMEM;
+				}
+			}
+		}
 
-		chunk = chunk->next;
+		size_t left = len;
+		size_t foffs = (size_t)offs;
+		size_t boffs = 0;
+		while (left > 0) {
+			void *chunk = o->chunks[DUMMYFS_CHUNKIDX(foffs)];
+			size_t chunkoffs = foffs % DUMMYFS_CHUNKSZ;
+			size_t cpylen = ((DUMMYFS_CHUNKSZ - chunkoffs) < left) ? (DUMMYFS_CHUNKSZ - chunkoffs) : left;
 
-	} while (len && chunk != o->chunks);
+			memcpy((char *)chunk + chunkoffs, (const char *)buff + boffs, cpylen);
 
-	o->atime = time(NULL);
-	object_unlock(fs, o);
-	object_put(fs, o);
+			left -= cpylen;
+			boffs += cpylen;
+			foffs += cpylen;
+			cnt += cpylen;
+		}
+	}
 
-	return ret;
+	return (int)cnt; /* FIXME: Should be ssize_t */
 }
 
 
 int dummyfs_write(void *ctx, oid_t *oid, offs_t offs, const char *buff, size_t len)
 {
+	TRACE;
 	dummyfs_t *fs = (dummyfs_t *)ctx;
-	dummyfs_object_t *o;
-	int ret = EOK;
 
-	o = object_get(fs, oid->id);
+	mutexLock(fs->mutex);
+	dummyfs_object_t *o = dummyfs_object_get(fs, oid);
 	if (o == NULL) {
+		mutexUnlock(fs->mutex);
 		return -EINVAL;
 	}
+	int ret = _dummyfs_writeObject(fs, o, offs, buff, len);
+	dummyfs_object_put(fs, o);
+	mutexUnlock(fs->mutex);
 
-	if (!S_ISREG(o->mode)) {
-		object_put(fs, o);
-		return -EINVAL;
-	}
-
-	object_lock(fs, o);
-
-	ret = dummyfs_write_internal(fs, o, offs, buff, len);
-
-	object_unlock(fs, o);
-	object_put(fs, o);
-
-	return ret;
+	return (ret == -ENOMEM) ? -ENOSPC : ret;
 }
 
 
 int dummyfs_createMapped(void *ctx, oid_t *dir, const char *name, void *addr, size_t size, oid_t *oid)
 {
+	TRACE;
 	dummyfs_t *fs = (dummyfs_t *)ctx;
-	dummyfs_object_t *o;
-	dummyfs_chunk_t *chunk;
-
-	if (dummyfs_create(ctx, dir, name, oid, 0755, otFile, NULL) != 0)
-		return -ENOMEM;
 
-	if ((o = object_get(fs, oid->id)) == NULL) {
-		dummyfs_destroy(ctx, oid);
-		return -EINVAL;
-	}
-
-	if ((chunk = malloc(sizeof(dummyfs_chunk_t))) == NULL) {
-		object_put(fs, o);
-		dummyfs_destroy(ctx, oid);
+	mutexLock(fs->mutex);
+	if (_dummyfs_create(fs, dir, name, oid, 0755, otFile, NULL) != 0) {
```

Comment:
return value is lost it's not always enomem

Reply:
Done